### ISSUE TITLE
RFC: Pipeline Configuration Redesign (pivot.yaml + Pipeline class)

### ISSUE THREAD

**Issue by sjawhar** [sjawhar/pivot#65]
## Summary

This RFC proposes a fundamental redesign of how pipelines are defined and discovered in Pivot. The goal is to support declarative configuration (`pivot.yaml`), programmatic composition (`Pipeline` class), and CLI commands that work without requiring Python imports.

## Problem Statement

Pivot has fundamental design limitations:

1. **Global registry pattern** - `@stage` decorator registers globally at import time
2. **CLI requires Python import** - Commands like `pivot metrics show` fail without registered stages
3. **No multiple DAGs** - Can't compose different pipelines from shared stages
4. **Boilerplate for DVC migration** - Importing DVC pipelines requires significant wrapper code

### 1. pivot.yaml References Undecorated Functions

```yaml
# pivot.yaml
stages:
  preprocess:
    python: mymodule.preprocess    # Plain function, no @stage decorator
    deps: [data/raw.csv]
    outs: [data/clean.csv]

  train:
    python: mymodule.train
    deps: [data/clean.csv]
    outs: [models/model.pkl]
    metrics:
      - metrics/train.json: {cache: false}
    params:
      learning_rate: 0.05           # Override class default
    mutex: [gpu]
```

```python
# mymodule.py - Plain functions, no decorators
import pydantic

class TrainParams(pydantic.BaseModel):
    learning_rate: float = 0.01
    epochs: int = 100

def preprocess():
    ...

def train(params: TrainParams):    # Type hint declares params class
    print(params.learning_rate)     # Will be 0.05 from pivot.yaml
    ...
```

### 2. Params Resolution Hierarchy

The params class is **introspected from the function signature** - no need to reference it in YAML.

```
1. Pydantic class defaults     →  learning_rate: 0.01
2. pivot.yaml overrides        →  learning_rate: 0.05
3. params.yaml overrides       →  learning_rate: 0.1  (final)
```

### 3. Pipeline Class Alternative (Programmatic)

```python
# pipeline.py
from mymodule import preprocess, train
from pivot import Pipeline

pipeline = Pipeline()
pipeline.add_stage(preprocess, deps=['data/raw.csv'], outs=['data/clean.csv'])
pipeline.add_stage(train, deps=['data/clean.csv'], outs=['models/model.pkl'],
                   params={'learning_rate': 0.05})  # Dict → introspect class from signature
```

Both approaches use the same mechanism: dynamic registration of plain functions.

---

### Dimension Format: List OR Dict

```yaml
matrix:
  # List: simple, key = value, no overrides
  dataset: [swe_bench, humaneval]

  # Dict: key with Hydra-style overrides
  model:
    bert:
      params:
        hidden_size: 768
    gpt:
      deps+: [tokenizer/gpt.json]    # Append to deps
      params:
        hidden_size: 1024
```

### Override Operations

| Syntax | Operation | Example |
|--------|-----------|---------|
| `key: value` | Replace | `deps: [only_this.csv]` |
| `key+: [items]` | Append to list | `deps+: [extra.csv]` |
| `key: {nested: val}` | Deep merge dict | `params: {lr: 0.05}` |

### Name Template via Stage Key

```yaml
stages:
  # Auto-naming: train@bert_swe_bench, train@gpt_humaneval, etc.
  train:
    python: mymodule.train
    matrix:
      model: [bert, gpt]
      dataset: [swe_bench, humaneval]
    deps: [data/${dataset}.csv]
    outs: [models/${model}_${dataset}.pkl]

  # Custom naming: eval@bert_on_swe, eval@gpt_on_human, etc.
  eval@{model}_on_{dataset}:
    python: mymodule.evaluate
    matrix:
      model: [bert, gpt]
      dataset: [swe, human]
    deps: [models/${model}_${dataset}.pkl]
    outs: [results/${model}_${dataset}.json]
```

### Validation Rules

- `@` in stage key without `matrix` → Error
- `matrix` without `@` → Auto-generate: `{name}@{dim1}_{dim2}_...`
- `@{dim}` template missing a dimension → Error (prevents duplicate names)

### Python Escape Hatch

```yaml
stages:
  complex:
    python: mymodule.complex_train
    variants: mymodule.generate_variants   # Returns list of variant dicts
```

---

## Manifest for CLI Discoverability

After `pivot run`, generate `.pivot/manifest.yaml`:

```yaml
# .pivot/manifest.yaml (auto-generated)
stages:
  preprocess:
    deps: [data/raw.csv]
    outs: [{path: data/clean.csv, type: Out}]
  train:
    deps: [data/clean.csv]
    outs:
      - {path: models/model.pkl, type: Out}
      - {path: metrics/train.json, type: Metric}
```

CLI commands read manifest for output discovery - no Python import needed.

---

## Complete pivot.yaml Schema

```yaml
stages:
  # Stage key: "name" or "name@{dim1}_{dim2}_..."
  <stage_name>[@{template}]:
    python: <module.function>           # Required: import path

    # Dependencies and outputs
    deps: [<path>, ...]
    outs: [<path>, ...]
    metrics:
      - <path>: {cache: false}
    plots:
      - <path>: {x: <col>, y: <col>}

    # Params (override class defaults)
    params:
      <field>: <value>

    # Execution
    mutex: [<group>, ...]
    cwd: <path>

    # Matrix expansion (optional)
    matrix:
      <dimension>: [<val>, ...]         # List form
      <dimension>:                       # Dict form
        <key>: {}                        # No overrides
        <key>:                           # With overrides
          deps+: [...]                   # Append
          params: {...}                  # Deep merge

    # OR: Python escape hatch
    variants: <module.function>
```

---

### Phase 1: Pipeline Class Foundation
- New `src/pivot/pipeline.py` with Pipeline class
- Modify `registry.py` to support dict params (introspect class from signature)

### Phase 2: pivot.yaml Parser
- New `src/pivot/pipeline_config.py` for YAML parsing and matrix expansion

### Phase 3: Matrix Expansion
- Cartesian product generation
- Variable interpolation (`${dim}`)
- Override application (append/merge)

### Phase 4: Manifest Generation
- New `src/pivot/manifest.py`
- Generate manifest after successful runs

### Phase 5: CLI Auto-Discovery
- Auto-discover `pivot.yaml` or `pipeline.py`
- Update CLI to use manifest for metrics/plots commands

### Phase 6: Fix Issue #62
- CLI commands work with explicit file targets via manifest

---

## Success Criteria

1. **pivot.yaml works**: Define pipeline in YAML, run with `pivot run`
2. **Pipeline class works**: Compose pipelines programmatically
3. **Params from signature**: No params class reference needed in YAML
4. **Matrix expansion**: List/dict dimensions with Hydra-style overrides
5. **CLI without import**: `pivot metrics show` works via manifest
6. **Backward compat**: Existing `@stage` decorator unchanged

---

## Related Issues

- #62 - CLI commands ignore explicit file targets when no stages registered
